docs/source/daml/reference/contract-keys.rst

```diff
@@ -0,0 +1,83 @@
+.. Copyright (c) 2019 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+.. SPDX-License-Identifier: Apache-2.0
+
+Contract keys
+#############
+
+Contract keys are an optional addition to templates. They let you specify a way of uniquely identifying instances of that contract: that can be a single parameter to the template or a combination of parameters (eg, in a tuple).
+
+You can use contract keys to stably refer to a contract, even through iterations of instances of it.
+
+Here's an example of setting up a contract key for a bank account, to act as a bank account ID:
+
+.. literalinclude:: ../code-snippets/ContractKeys.daml
+   :language: daml
+   :start-after: -- start contract key setup snippet
+   :end-before: -- end contract key setup snippet
+
+What can be a contract key
+**************************
+
+What you can specify is pretty flexible - you will probably want to set up a data type for your key, which is either a tuple or a record.
+
+However, the contract key *must* include every party that you specify as a ``maintainer``. For example, with ``maintainer bank``, ``key (bank, number) : AccountKey`` is valid; but wouldn't be if you removed ``bank`` from the key.
+
+Specifying maintainers
+**********************
+
+If you specify a contract key for a template, you must also specify ``maintainer``(s), in a similar way to specifying signatories or observers.
+
+Maintainers are the parties that know about all of the keys that they are party to, and so it's their responsibility to guarantee uniqueness of contract keys. The maintainers **must** be signatories or observers of the contract.
+
+When you're writing DAML models, the maintainers only matter since they affect authorization -- much like signatories and observers. You don't need to do anything to "maintain" the keys.
+
+Checking of the keys is done automatically at execution time, by the DAML exeuction engine: if someone tries to create a new contract that duplicates an existing contract key, the execution engine will cause that creation to fail. 
+
+Contract keys functions
+***********************
+
+Contract keys introduce several new functions.
+
+``fetchByKey``
```

Comment:
We need to explain the `@Account` part from the example above. This is about ambiguous types. I'll explain the details in person tomorrow.
Same for `lookupByKey` and `exerciseByKey`.